packages/application-shell/src/components/navbar/use-navbar-state-manager.ts

```diff
@@ -0,0 +1,197 @@
+            // concatenated and rendered the same way.
+            if (!app.navbarMenu) return;
+            return {
+              key: app.navbarMenu.key,
+              uriPath: app.navbarMenu.uriPath,
+              labelAllLocales: app.navbarMenu.labelAllLocales || [],
+              icon: app.navbarMenu.icon,
+              featureToggle: app.navbarMenu.featureToggle,
+              permissions: app.navbarMenu.permissions as string[],
+              submenu: (app.navbarMenu.submenu || []).map(menu => ({
+                key: menu.key,
+                uriPath: menu.uriPath,
+                labelAllLocales: menu.labelAllLocales || [],
+                featureToggle: menu.featureToggle,
+                permissions: menu.permissions as string[],
+                menuVisibility: undefined,
+                actionRights: undefined,
+                dataFences: undefined,
+              })),
+              menuVisibility: undefined,
+              actionRights: undefined,
+              dataFences: undefined,
+              shouldRenderDivider: false,
+            };
+          })
+          .filter(nonNullable)
+      : [];
+  const cachedIsForcedMenuOpen = window.localStorage.getItem(
+    STORAGE_KEYS.IS_FORCED_MENU_OPEN
+  );
+  const isForcedMenuOpen =
+    typeof cachedIsForcedMenuOpen === 'string'
+      ? cachedIsForcedMenuOpen === 'true'
+      : null;
+
+  const [activeItemIndex, setActiveItemIndex] = React.useState<string | null>(
+    null
+  );
+  const [isExpanderVisible, setIsExpanderVisible] = React.useState(true);
+  const [isMenuOpen, setIsMenuOpen] = React.useState(false);
+
+  const checkSize = React.useCallback(() => {
+    const shouldOpen = window.innerWidth > 1024;
+    const canExpandMenu = window.innerWidth > 918;
+
+    // If the screen is small, we should always keep the menu closed,
+    // no matter the settings.
+    if (!canExpandMenu) {
+      if (isMenuOpen || isExpanderVisible) {
+        // and resets the state to avoid conflicts
+        setIsMenuOpen(false);
+        setIsExpanderVisible(false);
+        setActiveItemIndex(null);
+      }
+    } else if (canExpandMenu && isExpanderVisible !== true)
+      setIsExpanderVisible(true);
+    else if (isNil(isForcedMenuOpen) && isMenuOpen !== shouldOpen) {
+      // User has no settings yet (this.props.isForcedMenuOpen === null)
+      // We check the viewport size and:
+      // - if screen is big, we open the menu
+      // - if screen is small we close it
+      setIsMenuOpen(shouldOpen);
+      setIsExpanderVisible(true);
+    } else if (!isNil(isForcedMenuOpen) && isMenuOpen !== isForcedMenuOpen) {
+      // User has setting, we should use that and ignore the screen size.
+      // Note: if viewport size is small, we should ignore the user settings.
+      setIsMenuOpen(isForcedMenuOpen);
+      setIsExpanderVisible(true);
+    }
+  }, [isExpanderVisible, isMenuOpen, isForcedMenuOpen]);
```

Comment:
This is still mad logic to me 😄. But it's good to move it over for now and then maybe partially redo the thing.